### PR TITLE
Slang

### PR DESCRIPTION
**Pull request by Janther**
This is a proposal for a version 2 of this plugin.

It has been rewritten leveraging the power of [Nomic Foundation's Slang](https://github.com/NomicFoundation/slang) to provide a much more scalable and better supported parser.

While we are still in beta we have disabled browser support until we can make sure we can fully provide this functionality.

The coverage has been reduced as well since `slang`'s AST is much more detailed than `solidity-parser`'s and while all our tests still pass, they don't provide every possible scenario.

Using with the old parser is still possible but a **deprecated** warning will be logged.

### CODE REVIEW

File: src/slang-nodes/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+  | StructMembers
+  | TupleDeconstructionElements
+  | TupleValues
+  | UnicodeStringLiterals
+  | UnnamedFunctionAttributes
+  | UsingDeconstructionSymbols
+  | VersionExpressionSet
+  | VersionExpressionSets
+  | VersionSpecifiers
+  | YulArguments
```
Comment (OmarTawfik) [inline src/slang-nodes/index.d.ts:285]: Note to self: we should probably export these types automatically from our API for each enum/its variants.

File: src/slang-printers/types.d.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import type { Doc } from 'prettier';
+
+interface PrintSeparatedOptions {
+  firstSeparator?: Doc;
+  separator?: Doc;
+  lastSeparator?: Doc;
```
Comment (OmarTawfik) [inline src/slang-printers/types.d.ts:6]: question: is it ever legal to have a "lastSeparator" in Solidity? my understanding is that it is always a parse error.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  Printer,
+  RequiredOptions,
+  SupportLanguage
+} from 'prettier';
+import type { AstNode } from './slang-nodes';
+
+const parserName = 'slang';
+const astFormat = 'slang-ast';
+
+// https://prettier.io/docs/en/plugins.html#languages
```
Comment (OmarTawfik) [inline src/index.ts:24]: is `slang-ast` the public user-facing "parser" name that uses are going to specify in their config?
In that case, what do you think about `slang-solidity` instead?

1. "ast" is a feature name, not tied to the parser, and might indicate otherwise.
2. allows supporting a `slang-yul` parser in the future, if it was supported: https://docs.soliditylang.org/en/latest/yul.html#stand-alone-usage

**Review by OmarTawfik — COMMENTED** [review]
I did an initial pass and left a couple of comments. Thanks for working on this. Exciting stuff!

File: src/slang-printers/types.d.ts (hunk 2 above)
Comment (Janther) [reply src/slang-printers/types.d.ts:6]: hehe welcome @OmarTawfik.

`lastSeparator` here refers to what happens just after the list is printed:

```Solidity
struct A {
    uint a; // firstSeparator is new line
    uint b;
    uint c; // lastSeparator would be a new line
}

assembly {
    let
        long, // firstSeparator is new line
        list,
        of,
        variable,
        names = f() // last separator would be empty string before the equal.
}
```

In most cases `lastSeparator` matches `firstSeparator` but in some cases it does not and it's necessary the extra parameter.

If you want to propose a better name to not have the `separator` word I'd happily update the PR with that.

File: src/slang-nodes/index.d.ts (hunk 1 above)
Comment (Janther) [reply src/slang-nodes/index.d.ts:285]: this is a list of object with `items` but the list of variants could also be helpful

Comment (Janther) [reply src/slang-nodes/index.d.ts:285]: to avoid confusion, I renamed it from `HubNode` to `CollectionNode`

Comment (OmarTawfik) [reply src/slang-nodes/index.d.ts:285]: I see. Would that still help in cases like `StatementVariant`?

File: src/slang-printers/types.d.ts (hunk 2 above)
Comment (OmarTawfik) [reply src/slang-printers/types.d.ts:6]: Thanks for explaining! I misunderstood how it is used. Up to you in this case.

File: README.md [hunk 4]
```diff
@@ -85,7 +91,7 @@ import solidityPlugin from 'prettier-plugin-solidity/standalone';
 
 async function format(code) {
   return await prettier.format(code, {
-    parser: "solidity-parse",
+    parser: "slang",
```
Comment (OmarTawfik) [inline README.md:94]: If `slang` is the user-facing parser name, I wonder what do you think about `slang-solidity` instead?

1. Is the parser name is "shared" between all prettier plugins, so it has to be unique, so it doesn't collide with another plugin/language?
2. I wonder if you are ever planning on supporting [stand-alone assembly files](https://docs.soliditylang.org/en/latest/yul.html#stand-alone-usage
) in the future, and therefore, you might want to introduce a `slang-yul` parser as well?

On the other hand, maybe this is overengineering/overthinking, and we can just leave it up to the user config, or the plugin can just auto-detect that from the file extension.
Just thinking out loud.

Comment (Janther) [reply README.md:94]: `slang-solidity` is a good idea.

Adding support for `.yul` is very straight forward as we just serve 2 parsers `slang-solidity` that expects a `SourceUnit` and `slang-yul` that expects a `YulBlock`.

Any thoughts on this @fvictorio?

File: src/slang-nodes/index.d.ts (hunk 1 above)
Comment (Janther) [reply src/slang-nodes/index.d.ts]: instead of listing them I just pushed a commit that automatically creates this type

File: src/slang-nodes/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+    operator: string;
+    rightOperand: Expression;
+  }
+>;
+
+export type AstNode =
+  | StrictAstNode
+  | Comment
+  | Identifier
+  | YulIdentifier
```
Comment (Janther) [inline src/slang-nodes/index.d.ts:472]: @OmarTawfik It would be beneficial if `Identifier` and `YulIdentifier` were in the types.
For example:
```TypeScript
export declare class EnumDefinition {
    readonly cst: NonterminalNode;
    private readonly fetch;
    constructor(cst: NonterminalNode);
    get enumKeyword(): TerminalNode;
    get name(): TerminalNode;
    get openBrace(): TerminalNode;
    get members(): EnumMembers;
    get closeBrace(): TerminalNode;
}
```
I care about getting the location information of `name` but not from `enumKeyword`, `openBrace` and `closeBrace`.
I went through every `TerminalNode` in the AST and plucked out the `Identifier`s and `YulIdentifier`s but if you make a change in the AST I have no Type security that will trigger an error on my side if I need to adapt. So I would appreciate if it would be possible to deliver something like:

```TypeScript
export declare class EnumDefinition {
    readonly cst: NonterminalNode;
    private readonly fetch;
    constructor(cst: NonterminalNode);
    get enumKeyword(): TerminalNode;
    get name(): Identifier;
    get openBrace(): TerminalNode;
    get members(): EnumMembers;
    get closeBrace(): TerminalNode;
}
```

Comment (OmarTawfik) [reply src/slang-nodes/index.d.ts:472]: Thanks for the suggestion!

> but if you make a change in the AST I have no Type security that will trigger an error on my side

To make sure I understand you correctly, do you mean `getNodeMetadata().offsets`?
Is that because you are storing `offsets` as an array, and accessing it by a numeric index? Would it help if it was changed to be an object/dict with the field name as keys? for example `offsets.name` here won't break if the position of `name` was changed relative to its siblings.

> It would be beneficial if Identifier and YulIdentifier were in the types.

How would that help in this case? if they were non-terminal nodes? how would this be tracked otherwise?

One other way you can use Slang API if you want, so that you don't have to do the manual tracking of offsets yourself, is to use cursors.
For example, in `getNodeMetadata()`, instead of iterating over `ast.cst.children()`, you can `ast.cst.createCursor()`, and then use `cursor.goToNextSibling()` to iterate, and `cursor.textOffset()` or `cursor.textRange()` will be updated.

Comment (Janther) [reply src/slang-nodes/index.d.ts:472]: 🤔 using a Map with the cst as the key could really improve this aspect of the implementation.
it would even render the passing of an offset moot since the we are already passing the AST to the constructors.
I'll experiment with this and see how it behaves.

File: README.md (hunk 4 above)
Comment (fvictorio) [reply README.md]: I'd say let's open an issue about this and not include it here, since this change is massive as is.

File: src/slang-nodes/index.d.ts (hunk 5 above)
Comment (OmarTawfik) [reply src/slang-nodes/index.d.ts:472]: Thanks for pointing out. The `id` already exists in the Rust API, and I would be happy to add it in TypeScript. It should be a simple change.
When I'm done with WASM, I will also be able to expose the field names themselves as a strongly-typed `EdgeLabel` enum, so you would be able to refer to and persist the name of `get name(): Identifier ` field as `EdgeLabel.Name`.

**Comment by Janther** [comment]
> I didn't do an in-depth review of the PR; this is just the result of manually testing it and skimming the code for a couple of hours. If there are things you would like me to review in depth, please point them out to me, because I don't think I'll be able to do a line-by-line review of the whole PR.

The only thing that pops up is the `slang-comments` section which is new.

> ### `Comments.sol` tests
> These tests are only run with the ANTLR parser, and the `run-format-test.js` file has a comment that says `// TODO: finish Comments`. Does this mean that there's something pending around comments when using the Slang parser?

These were some extremely edge cases with the positioning of comments where honestly comments should not go. I pushed a commit making these scenarios, not the same as with the antlr parser but closer to what prettier does with JS in a similar situation.

> And, as an aside, all the Slang-related code in `run-format-test.js` doesn't seem to make sense? `shouldTestSlang` will always return false, so all the code used in that scenario seems like dead code.

yeah this was dead code (just removed it) that was helpful when I was slowly implementing new printers and had a proper list of tests that needed to be tested with both parsers.

> ### Defaulting to the latest supported version
> I mentioned this already before, but defaulting to the latest Solidity version supported by slang won't work for projects with multiple versions of Solidity, which is a common thing for projects that have been around for some years.
> 
> I don't know what's the right solution here. I kind of feel that Slang should export some helper utility to let you infer the version to use for a given code (by getting the pragmas doing what [solidity-analyzer](https://github.com/NomicFoundation/solidity-analyzer) does, and then returning the latest supported version that matches the pragmas). But there might be better solutions.

This is planned but not in the scope of this PR.

> ### Other comments
> * There are some commented out lines in `CI.yml`

While we slang doesn't support WASM, the build process is broken and the CI step for browser should be skipped.
Once they start supporting it, the CI comes right back.

> * The `options.parentParser` trick used in the `SourceUnit` nodes is, if I remember correctly, done to avoid printing empty trailing lines in Solidity snippets within a markdown file. Can we add a comment explaining that? Otherwise it's a bit obscure.

Done.

> * Out of scope for the Slang refactor, but I think we should use `tsx` instead of `ts-node`.

Will look into this.

**Review by fvictorio — APPROVED** [review]
